src/spatial/OctNode.java

```diff
@@ -0,0 +1,182 @@
+package spatial;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.lwjgl.util.vector.Matrix4f;
+import org.lwjgl.util.vector.Vector4f;
+
+import renderer.Renderable;
+import renderer.model.Cullable;
+import util.MathUtils;
+import util.Plane;
+
+/***
+ * TODO: Needs an update function
+ * @author Max
+ *
+ */
+public class OctNode implements Renderable {
+	private static final int CAPACITY = 8;	
+	private static final int DIM = 3;
+	private OctNode[] children;
+	private int count;
+	private List<Cullable> objects;
+	private boolean isBound;
+	
+	// Max Depth of Subtrees
+	private final int maxDepth;
+	
+	// Coordinates and half width of node
+	private float x;
+	private float y;
+	private float z;
+	private float[] pos;
+	
+	// Half the width of the current volume (Note: seems that canonically, this is half the width of the child)
+	private float halfWidth;
+	
+	public OctNode(float x, float y, float z, float halfWidth, int maxDepth) {
+		this.x = x;
+		this.y = y;
+		this.z = z;
+		this.pos = new float[] {x, y, z};
+		this.halfWidth = halfWidth;
+		this.children = new OctNode[CAPACITY];
+		this.maxDepth = maxDepth;
+		this.count = 0;
+		this.isBound = false;
+		this.objects = new ArrayList<>();
+		
+		// Initialize the children to null
+		for(int i = 0; i < CAPACITY; ++i)
+			this.children[i] = null;
+	}
+
+	private static OctNode createChild(OctNode parent, int index) {
+		float halfWidth = parent.halfWidth / 2;
+		float x = ((index & 1) > 0 ? halfWidth : -halfWidth);
+		float y = ((index & 2) > 0 ? halfWidth : -halfWidth);
+		float z = ((index & 4) > 0 ? halfWidth : -halfWidth);
+		return new OctNode(parent.x + x, parent.y + y, parent.z + z, halfWidth, parent.maxDepth - 1);
+	}
+	
+	public void insert(Cullable obj) {
+		boolean straddle = false;
+		int index = 0;
+		count++;
+		
+		for(int i = 0; i < DIM; ++i) {
+			float delta = obj.getCentre(i) - pos[i];
+			straddle |= Math.abs(delta) < (halfWidth / 2 + obj.getWidth(i) / 2);
+			
+			// Map the position to an octant
+			if(delta > 0)
+				index |= (1 << i);
+		}
+		
+		System.out.println("Max depth: " + maxDepth);
+		// If the object straddles a geometric boundary for the node, or if the max depth is reached add to this level
+		if(straddle || maxDepth == 0) {

```

Comment:
slight optimization, we can put:
```java
if (maxDepth == 0) {
    objects.add(obj);
    return;
}
```
at the top of this function, so we dont have to do all the extra computations in that for loop.